src/main/java/io/aiven/kafka/connect/common/grouper/ParquetTopicPartitionRecordGrouper.java

```diff
@@ -0,0 +1,116 @@
+
+import io.aiven.kafka.connect.common.config.TimestampSource;
+import io.aiven.kafka.connect.common.templating.Template;
+
+class ParquetTopicPartitionRecordGrouper extends TopicPartitionRecordGrouper {
+
+    private final SchemaBasedRotator schemaBasedRotator = new SchemaBasedRotator();
+
+    ParquetTopicPartitionRecordGrouper(final Template filenameTemplate,
+                                       final Integer maxRecordsPerFile,
+                                       final TimestampSource tsSource) {
+        super(filenameTemplate, maxRecordsPerFile, tsSource);
+    }
+
+    @Override
+    protected String resolveRecordKeyFor(final SinkRecord record) {
+        if (schemaBasedRotator.rotate(record)) {
+            return generateNewRecordKey(record);
+        } else {
+            return super.resolveRecordKeyFor(record);
+        }
+    }
+
+    @Override
+    public void clear() {
+        schemaBasedRotator.clear();
+        super.clear();
+    }
+
+    private static final class SchemaBasedRotator implements Rotator<SinkRecord> {
+
+        private final Map<TopicPartition, KeyValueSchema> schemaVersions = new HashMap<>();
+
+        @Override
+        public boolean rotate(final SinkRecord record) {
+            if (Objects.isNull(record.valueSchema()) || Objects.isNull(record.keySchema())) {
+                throw new SchemaProjectorException("Record must have schemas for key and value");
+            }
+            final var tp = new TopicPartition(record.topic(), record.kafkaPartition());
+            final var keyValueVersion =
+                    schemaVersions.computeIfAbsent(tp, ignored -> new KeyValueSchema(
+                            record.keySchema(),
+                            record.valueSchema()));
+            final var schemaChanged =
+                    !keyValueVersion.keySchema.equals(record.keySchema())
+                            || !keyValueVersion.valueSchema.equals(record.valueSchema());
+            if (schemaChanged) {
+                schemaVersions.put(tp,
+                        new KeyValueSchema(record.keySchema(), record.valueSchema())
+                );
+            }
+            return schemaChanged;
+        }
+
+        private static class KeyValueSchema {
+
+            final Schema keySchema;
+
+            final Schema valueSchema;
+
+            KeyValueSchema(final Schema keySchema, final Schema valueSchema) {
+                this.keySchema = keySchema;
+                this.valueSchema = valueSchema;
+            }
+
+            @Override
+            public boolean equals(final Object o) {
+                if (this == o) {
+                    return true;
+                }
+                if (o == null || getClass() != o.getClass()) {
+                    return false;
+                }
+                final var that = (KeyValueSchema) o;
+                return keySchema.equals(that.keySchema) && valueSchema.equals(that.valueSchema);
```

Comment:
Consider using `Objects.equals`, it doesn't seem guaranteed that `keySchema` and `valueSchema` are non-null.

Reply:
In this case it is ok, this class is private and schemas appear after been checking on null